gengokit/gengokit.go

```diff
@@ -86,37 +118,103 @@ func getProtoService(dt deftree.Deftree) (*deftree.ProtoService, error) {
+		log.WithField("Template file", templFP).Debug("Skipping rendering non-buildable partial template")
+		return nil, nil
 	}
-}
 
-// getFileContentByName searches though a []truss.NamedReadWriter and returns the
-// contents of the file with the name n
-func getFileByName(n string, files []truss.NamedReadWriter) io.Reader {
-	for _, f := range files {
-		if f.Name() == n {
-			return f
+	var genCode io.Reader
+	var err error
+
+	// Get the actual path to the file rather than the template file path
+	actualFP := templatePathToActual(templFP, te.PackageName)
+
+	// Map of template paths to template rendering functions
+	renderFuncs := map[string]func(io.Reader, *templateExecutor) (io.Reader, error){
+		"NAME-service/handlers/server/server_handler.gotemplate": updateServerMethods,
+		"NAME-service/handlers/client/client_handler.gotemplate": updateClientMethods,
+	}
+
+	// If we are rendering a template with a renderFunc and the file existed
+	// previously then use that function
+	if renderFuncs[templFP] != nil {
+		file := prevGenMap[actualFP]
+		if file != nil {
+			genCode, err = renderFuncs[templFP](file, te)
 		}
 	}
-	return nil
+
+	if err != nil {
+		return nil, errors.Wrap(err, "could not render template")
+	}
+
+	// if no code has been generated just apply the template
+	if genCode == nil {
+		genCode, err = applyTemplateFromPath(templFP, te)
+	}
+
+	if err != nil {
+		return nil, errors.Wrap(err, "could not render template")
+	}
+
+	// Turn code buffer into string and format it
+	codeBytes, err := ioutil.ReadAll(genCode)
+	if err != nil {
+		return nil, err
+	}
+
+	// ignore error as we want to write the code either way to inspect after
+	// writing to disk
+	formattedCode, _ := formatCode(codeBytes)
+
+	var resp truss.SimpleFile
+
+	// Set the path to the file and write the code to the file
+	resp.Path = actualFP
+	resp.Write(formattedCode)
+
+	return &resp, nil
+}
+
+// templatePathToActual accepts a templateFilePath and the packageName of the
+// service and returns what the relative file path of what should be written to
+// disk
+func templatePathToActual(templFilePath, packageName string) string {
+	// Switch "NAME" in path with packageName.
+	//i.e. for packageName = addsvc; /NAME-service/NAME-server -> /addsvc-service/addsvc-server
+	actual := strings.Replace(templFilePath, "NAME", packageName, -1)
+
+	actual = strings.TrimSuffix(actual, "template")
+
+	return actual
 }
 
-// updateServerMethods will update the functions within an existing server_handler.go
-// file so it contains all the functions within svcFuncs
-func (g *generator) updateServerMethods(svcHandler io.Reader, svcFuncs []string) (outCode *bytes.Buffer, err error) {
+// filesTofilepathMap accepts a slice of truss.NamedReadWriters and returns a map of filepaths
+// as strings to files as io.Reader
+func filesToFilepathMap(previousFiles []truss.NamedReadWriter) (pathToFile map[string]io.Reader) {
+	pathToFile = make(map[string]io.Reader)
```

Comment:
When you know how many maps an element will hold, you can pre-initialize it. Probably not a big deal here since it will probably all fit in the initial initialization (and performance is not going to be an issue with Truss, generally speaking), but this could be
```
pathToFile = make(map[string]io.Reader, len(previousFiles))
```
However, I would get rid of the named return parameter here. You're returning it by name, so you're not using it except to eliminate a colon in the initial assignment. Generally speaking, avoid named return parameters unless you have a compelling reason to use them. They're generally less readable than variables declared inline.